output/elasticsearch/watcher.go

```diff
@@ -25,67 +27,136 @@ type WatchReporter func(state WatchState, msg string)
 
 // ESHealthWatcher monitors the ES connection and notifies if a failure persists for more than failureInterval seconds
 type ESHealthWatcher struct {
-	reporter        WatchReporter
-	lastSuccess     time.Time
-	lastFailure     time.Time
-	didFail         bool
-	lastFailMessage string
-	failureInterval time.Duration
-	waitInterval    time.Duration
-
-	mut sync.Mutex
+	reporter          WatchReporter
+	callbackFailCount int32
+	didFail           bool
+	failureInterval   time.Duration
+	waitInterval      time.Duration
+	esClient          ConnectionMetrics
+	log               *logp.Logger
+}
+
+// ConnectionMetrics is an interface type for ES transport metrics
+type ConnectionMetrics interface {
+	Metrics() (elastictransport.Metrics, error)
+}
+
+//////////// Test helpers
+
+// sets the ticker channel used by the watcher. Used for tests.
+var monitoringTicker = createWatchTicker
+
+// default ticker for the health monitor, can by bypassed for tests
+func createWatchTicker(interval time.Duration) <-chan time.Time {
+	return time.NewTicker(interval).C
 }
 
-func newHealthWatcher(reporter WatchReporter, failureInterval time.Duration) ESHealthWatcher {
+// If set, is called every time the watcher loop returns to the start
+// This helps us avoid copious sleep() statements in the tests
+var watcherLoopResetChan chan struct{}
+
+////////////
+
+func newHealthWatcher(log *logp.Logger, reporter WatchReporter, failureInterval time.Duration, client ConnectionMetrics) ESHealthWatcher {
 	return ESHealthWatcher{
+		log:             log,
 		reporter:        reporter,
 		didFail:         false,
 		failureInterval: failureInterval,
-		waitInterval:    time.Second,
+		waitInterval:    time.Second * 5,
+		esClient:        client,
 	}
 }
 
-// Fail updates the time and error message of the last known failure
-func (hw *ESHealthWatcher) Fail(msg string) {
-	hw.mut.Lock()
-	defer hw.mut.Unlock()
-	hw.lastFailure = time.Now()
-	hw.lastFailMessage = msg
-}
+// Fail increments the count of callback failures
+// This helps us determine health status in cases where errors originate outside of the ES transport. JSON errors, etc
+func (hw *ESHealthWatcher) Fail() {
+	atomic.AddInt32(&hw.callbackFailCount, 1)
 
-// Success updates the last known successful ES call
-func (hw *ESHealthWatcher) Success() {
-	hw.mut.Lock()
-	defer hw.mut.Unlock()
-	hw.lastSuccess = time.Now()
 }
 
-// Watch starts a blocking loop and will report if there has been a failure for longer than a given period
+// Watch starts a blocking loop and will report if there has been a failure for longer than a given period.
+// This is a blocking call
 func (hw *ESHealthWatcher) Watch(ctx context.Context) {
+
+	// count of failures from the es transport
+	lastFail := 0
```

Comment:
Can we change some of these names to be more precise (e.g. it's kind of confusing when `lastFail` is an integer and `lastFailure` is a timestamp). Suggestions:
- `lastFail` -> `transportFailCount`
- `lastResponseFailRange` -> `statusFailCount`
- `lastResponseSuccessfulRange` -> `statusSuccessCount`